route: Validate the route destination address

liangwen12year opened the pull request:
'0.0.0.0/8' and its subnet should never be used as the destination address for the route, thus raise the invalid argument error during the route validation.

Signed-off-by: Wen Liang <liangwen12year@gmail.com>

liangwen12year committed:
route: Validate the route destination address

'0.0.0.0/8' and its subnet should never be used as the destination
address for the route, and it may be a typo of the default gateway,
thus raise the error during the route validation.

Signed-off-by: Wen Liang <liangwen12year@gmail.com>

cathay4t commented on rust/src/lib/route.rs:
```diff
@@ -78,6 +78,22 @@ impl Routes {
                     error!("{}", e);
                     return Err(e);
                 }
+                if !route.destination.as_deref().unwrap_or_default().is_empty() {
```
Please use `if let Some(dst) = route.destination.as_ref()` to save the extra check on `String::new().is_empty()`.

liangwen12year replied:
I updated. Thanks.

cathay4t commented on rust/src/lib/route.rs:
```diff
@@ -78,6 +78,22 @@ impl Routes {
+                            ErrorKind::InvalidArgument,
+                            format!(
+                                "0.0.0.0/8 and its subnet cannot be used as \
+                            the route destination: {:?}",
```
Normally we do not show debug output to user as error message.

Meanwhile, provide a hint to user that they might want to use default gateway `0.0.0.0/0`.

cathay4t reviewed (changes requested):
Besides change request, please fix CI.

cathay4t commented:
And test case is also required.

liangwen12year commented:
> And test case is also required.

Test case is added.

ffmancera commented on rust/src/lib/route.rs:
```diff
@@ -77,6 +78,24 @@ impl Routes {
                 }
+                if let Some(dst) = route.destination.as_ref() {
+                    if !is_ipv6_addr(dst.as_str()) {
+                        let net: Ipv4Net = dst.parse()?;
```
There is a problem here. This works only for addresses with a prefix. That means it will parse `192.0.2.0/24` correctly but it will fail on `192.0.2.0`. Checking ipnet docs you will find `Ipv4Net provides a [FromStr](https://doc.rust-lang.org/std/str/trait.FromStr.html) implementation for parsing network addresses represented in CIDR notation`

It needs to be fixed, thank you!

ffmancera commented on rust/src/lib/route.rs:
```diff
@@ -77,6 +78,33 @@ impl Routes {
                 }
+                if let Some(dst) = route.destination.as_ref() {
+                    let dst_str = dst.as_str();
+                    if !is_ipv6_addr(dst_str) {
+                        let mut is_invalid_dst = false;
+                        if dst_str == "0.0.0.0" {
+                            is_invalid_dst = true;
+                        } else if dst_str.contains('/') {
+                            let net: Ipv4Net = dst.parse()?;
+                            if Ok(net.network()) == "0.0.0.0".parse()
+                                && net.prefix_len() >= 8
+                            {
+                                is_invalid_dst = true;
+                            }
+                        }
+                        if is_invalid_dst {
+                            let e = NmstateError::new(
+                                ErrorKind::InvalidArgument,
+                                "0.0.0.0/8 and its subnet cannot be used as \
+                            the route destination, please use the default \
+                            gateway 0.0.0.0/0 instead"
+                                    .to_string(),
+                            );
+                            error!("{}", e);
+                            return Err(e);
+                        }
```
It seems a little bit complex IMO. I think you can skip the validation if the address is not in CIDR notation. In order to do that you can combine it with the `is_ipv6_addr` if statement. That means it will look something like that `if !is_ipv6_addr() && dst_str.contains('/')`. I think that is the right approach instead of having an extra variable to handle if the dst is invalid or not.

cathay4t commented on rust/src/lib/route.rs:
```diff
@@ -77,6 +78,33 @@ impl Routes {
+                            let e = NmstateError::new(
+                                ErrorKind::InvalidArgument,
+                                "0.0.0.0/8 and its subnet cannot be used as \
+                            the route destination, please use the default \
```
Please fix the indent.

liangwen12year replied:
fixed.

cathay4t reviewed (changes requested):
Using `ipnet::Ipv4Net` is overkill. Simple `split()` and compare first entry for `0.*` and second entry for `>=8`.

When we introduce something from crate, we are simplify the code, not make it worse.

cathay4t commented:
More detail:
 * `let ip_net: Vec<&str> = addr.split('/').collect()`
 * Use `Ipv4Addr::from_str(ip_net[0])` and `u8::parse(ip_net[1]).unwrap_or_default()`.
 * Compare `IPv4Addr.octets()[0] == 0` and `prefix >= 8`.

With this, we do not depend on ipnet which I intend to remove its depency from nmstate.

cathay4t commented on rust/src/lib/route.rs:
```diff
@@ -77,6 +79,34 @@ impl Routes {
+                        } else if dst_str.contains('/') {
+                            let ip_net: Vec<&str> =
+                                dst_str.split('/').collect();
+                            let ip_addr = Ipv4Addr::from_str(ip_net[0]);
+                            let prefix = ip_net[1].parse::<i32>().unwrap_or(32);
+                            if ip_addr?.octets()[0] == 0 && prefix >= 8 {
```
Normally, we do not place `?` in the middle of statement. Please try `let ip_addr = Ipv4Addr::from_str(ip_net[0])?;"

liangwen12year replied:
I updated. thanks.

cathay4t reviewed (changes requested):
Please place these check code into a dedicate function like `is_dst_valid() -> bool` with comment to this function explaining what you are validating.

In general, using `mut is_invalid_dst` as flag is a bad practice, considering this flag been modified or overrided after 100+ lines, developer could easily forgot the exact original purpose.

cathay4t commented on rust/src/lib/route.rs:
```diff
@@ -77,6 +79,34 @@ impl Routes {
+                            let ip_net: Vec<&str> =
+                                dst_str.split('/').collect();
+                            let ip_addr = Ipv4Addr::from_str(ip_net[0]);
+                            let prefix = ip_net[1].parse::<i32>().unwrap_or(32);
```
The `unwrap_or(32)` is wrong. What if I use `0.0.0.0/a`.

cathay4t replied:
BTW. The `ip_net[1]` will panic if the dst is `1.1.1.1`.

cathay4t replied:
Using `[]` is dangerous unless you really sure the index exists. I think `ip_net[0]` might cause problem on empty string `dst`, as a library, we should __never__ panic!

liangwen12year replied:
> BTW. The `ip_net[1]` will panic if the dst is `1.1.1.1`.

I think that the dst will not be `1.1.1.1`, because we the code will be called under the if condition `else if dst_str.contains('/') `.

liangwen12year replied:
> Using `[]` is dangerous unless you really sure the index exists. I think `ip_net[0]` might cause problem on empty string `dst`, as a library, we should **never** panic!

I am pretty sure that the index exists base on the if condition `else if dst_str.contains('/')`. The dst can not be empty string under the if condition.

cathay4t commented on rust/src/lib/route.rs:
```diff
@@ -402,3 +420,22 @@ fn flat_absent_route(
+fn is_route_dst_valid(dst_str: &str) -> Result<bool, AddrParseError> {
+    if dst_str.contains('/') {
+        let ip_net: Vec<&str> = dst_str.split('/').collect();
+        let ip_addr = Ipv4Addr::from_str(ip_net[0])?;
```
This will panic if `dst_str` is empty string.

cathay4t commented on rust/src/lib/route.rs:
```diff
@@ -402,3 +420,22 @@ fn flat_absent_route(
+    if dst_str.contains('/') {
+        let ip_net: Vec<&str> = dst_str.split('/').collect();
+        let ip_addr = Ipv4Addr::from_str(ip_net[0])?;
+        let prefix = ip_net[1].parse::<i32>().unwrap_or_default();
```
This means `0.0.0.0/f` is equal to `0.0.0.0/0` with your parser.

cathay4t commented on rust/src/lib/unit_tests/route.rs:
```diff
@@ -436,3 +436,33 @@ next-hop-address: "2001:db8:a:0000:000::1"
+    let result = routes.validate();
+    assert!(result.is_err());
+    assert_eq!(result.err().unwrap().kind(), ErrorKind::InvalidArgument);
+}
```
Please include more test on string like:

 * `0.0.0.0/f`
 * `0.0.0.0.0/0`
 * `0.0.0.0.0/7`
 * `0.0.0.0`

cathay4t commented on rust/src/lib/ifaces/sriov.rs:
```diff
@@ -79,11 +79,11 @@ impl SrIovConfig {
 
             if !vfs.is_empty() {
-                let total_vfs = self.total_vfs.unwrap_or(
+                let total_vfs = self.total_vfs.unwrap_or_else(|| {
```
This should in another patch

cathay4t commented on rust/src/lib/route.rs:
```diff
@@ -77,6 +79,22 @@ impl Routes {
+                if let Some(dst) = route.destination.as_ref() {
+                    let dst_str = dst.as_str();
+                    if !is_ipv6_addr(dst_str)
+                        && !is_route_dst_valid(dst_str).unwrap_or_default()
```
You have two layer of raising errors. And this `!foo().unwrap_or_default()` is hurting human eyes.

Try change to something like:


```rust
if let Some(dst) = route.destination.as_deref() {
    validate_route_dst(dst)?;
}


fn validate_route_dst(dst: &str) -> Result<(), NmstateError> {
    todo!()
}
```

liangwen12year committed:
rust/lib: Replace unwrap_or with unwrap_or_else

When passing the result of a function call, it is recommended to use
unwrap_or_else instead of unwrap_or.

https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call

Signed-off-by: Wen Liang <liangwen12year@gmail.com>

cathay4t commented on rust/src/lib/route.rs:
```diff
@@ -468,3 +473,72 @@ impl MergedRoutes {
+            } else if dst.contains('/') {
+                let ip_net: Vec<&str> = dst.split('/').collect();
+                let ip_addr = Ipv4Addr::from_str(ip_net[0])?;
+                let prefix = ip_net[1].parse::<i32>().unwrap_or_default();
+                if prefix == 0 && ip_net[1] != "0" {
```
Why not just:

```
let prefix = match ip_net[1].parse::<i32>() {
    Ok(p) => p,
    Err(e) => ...
}
```

cathay4t commented on rust/src/lib/route.rs:
```diff
@@ -468,3 +473,72 @@ impl MergedRoutes {
+                if prefix == 0 && ip_net[1] != "0" {
+                    let e = NmstateError::new(
+                        ErrorKind::InvalidArgument,
+                        "The prefix of the route destination network is \
```
When you raise errors, it is better to include the original invalid value, so user know which line to fix.

Something like `format!("The prefix of the route destination '{dst}' is invalid")`.

cathay4t commented on rust/src/lib/route.rs:
```diff
@@ -468,3 +473,72 @@ impl MergedRoutes {
+    if !is_ipv6_addr(dst) {
+        if dst.contains('.') {
+            let dot_count = dst.matches('.').count();
+            if dot_count != 3 {
```
This is not required. The `Ipv4Addr::from_str(ip_net[0])?;` can do the magic.

cathay4t commented on rust/src/lib/route.rs:
```diff
@@ -473,3 +478,64 @@ impl MergedRoutes {
+            ErrorKind::InvalidArgument,
+            format!(
+                "The route destination network '{dst}' is invalid and not \
+            in dotted decimal format"
```
The `Ipv4Addr::from_str()` handle this. You do not need to handle it.

cathay4t commented on rust/src/lib/route.rs:
```diff
@@ -473,3 +478,64 @@ impl MergedRoutes {
+fn validate_route_dst(dst: &str) -> Result<(), NmstateError> {
+    if !is_ipv6_addr(dst) {
+        if dst.contains('.') {
+            if dst.contains('/') {
```
You can do split regardless it contains `/` or not, you could save some lines of codes.

liangwen12year commented:
/retest-required

ffmancera committed:
routes: validate desired routes when creating MergedRoutes

This is the right place where we should validate the desired routes.
This call was missing and before there wasn't validation at all.

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>
